### ISSUE TITLE
Bug: server error when voting, unable to advance meeting

### ISSUE THREAD

**Issue by jordanh** [ParabolInc/parabol#5249]
## Issue - Bug

This user reported first reported #5248.

When investigating further, the user reported:

> We had several issues but I think the root cause was that we got stuck in the grouping phase
>  * Once we were finished with the Grouping Phase, I advanced the meeting to the Vote Phase. Only one person in the >. 
>  * meeting moved to the Vote Phase with me the rest were in the Grouping phase still.
>  * When the two of us tried to Vote we go a Server Error but if we refreshed our screen the vote would show up
>  * I could not advance to the discussion phase since no one could vote

The meeting ID is `5cVMwQLims`

One note: if attempt to revert the meeting to an incomplete state by removing the `endedAt` and `summarySentAt` fields from the `NewMeeting` record, server error is emitted to the client UI. I wasn't able to track down the actual server error.

### Acceptance Criteria (optional)
Bug triaged:
   - Triaged and fixed or new issue to fix opened with root cause
Triage is performed to find the root cause of the bug, timeboxed to ~1 hour. 

**Estimated effort:** 4 hours to triage. More if root cause is already identified.

**Comment by tianrunhe** [comment]
The meeting object was in a malformed state:

![Screen Shot 2021-08-10 at 4 40 37 PM](https://user-images.githubusercontent.com/1879975/129097825-0e077184-af35-409b-b4c5-c8b46b5b9da7.png)

Notice that the vote stage was written to the group phase and discuss stage was written to the vote phase.

**Comment by tianrunhe** [comment]
Right now there are 9 meetings in total that have a vote phase with stages that are not vote. 8 meetings in total that have discuss phases with 0 stages.

The earliest occurrence was on 10/09/2020. We released the feature for resetting to group phase on 10/07/2020 so it seemed correlated.

Next step: investigate on how reseting meeting phases could end up in misplaced stages.

**Comment by tianrunhe** [comment]
btw all malformed meetings are retrospective meetings, not surprisingly.

The linear structure of a retro meeting should look like, phase/stage pair: reflect/reflect -> group/group -> vote/vote -> discuss/discuss. In the malformed meeting examples I've seen:
* reflect/reflect -> group/vote -> vote/discuss -> discuss/empty
* reflect/reflect -> group/reflect -> vote/group -> discuss/vote
So it feels like an array index related issue? Like something that takes the subarray and then insert it at a wrong position.

**Comment by mattkrick** [comment]
looking through the code, the looping over old phases & using the new phases indexes feels error prone: https://github.com/ParabolInc/parabol/blob/db45e3b5a4c27e917dcc267d33175bae8a04970b/packages/server/graphql/mutations/resetMeetingToStage.ts#L88

for example, what if they start a meeting, then disable the check-in phase in their settings, then reset the meeting? we'd have an off by one error. 

not sure if this was the exact cause, but we've been meaning to rewrite that bit of code for some time now, so let's do it & see what we get. we should also lock this mutation down to only retro meetings since we have a lot of retro-only logic that could fail if they were jerks & called it with a different meeting type.

could you take a crack at cleaning this up? ideally we don't even create new meeting phases (since this only works with retros that reset to the group phase today). just find the vote phase, nuke the data. find the discuss phase, nuke all that stuff & put a dummy discuss stage back in. 

you can also use the `meetingId` index to clean up this logic: https://github.com/ParabolInc/parabol/blob/db45e3b5a4c27e917dcc267d33175bae8a04970b/packages/server/graphql/mutations/resetMeetingToStage.ts#L117

**Comment by tianrunhe** [comment]
Oh boy this is a tricky one! I can finally reproduce it consistently on my local. Here are the steps:

1. Open two tabs on retro meeting lobby page for the same team
2. On one tab, check "Include icebreaker" and starts the meeting. Let's call it meeting 1.
3. On the other tab, notice the "Include icebreaker" should reflect the checked state. Uncheck it though and starts another meeting. Call it meeting 2.
4. Now on meeting 1, go through reflect -> group -> vote
5. On meeting 1, go back to group phase and click "Edit Groups" and confirm.
6. Congratulations you have created a malformed meeting object! If you look at the DB you will see it. On client side, you can still go to the vote phase but voting will pop a "server error" toast.

**Comment by tianrunhe** [comment]
I think a broader issue that this bug reflects is that the `meetingSetting` object is shared among meetings. So technically if someone was fast enough to mutate that object **before** the start meeting mutation finishes, the new meeting could end up in a wrong state than settings user picked before starting: e.g., whether icebreaker is included, or even which template to use. The `meetingSetting` today is a team-level object that basically represents settings from **last** meeting. What we really need is a per-meeting scope setting. We might do the similar thing we do for poker template to "freeze" it. This looks like a lot of effort.

Another culprit for `resetMeetingToStage` is that it tries to be the generic mutation where you can reset to **any** stage. But we only need it to reset to group stage today. We can change that mutation to much smaller scope so we can bypass the `meetingSettings` thingy entirely.

@mattkrick which approach do you think work the best for the moment?
